File: leetcode/0121_max_profit.rs
struct Solution;

impl Solution {
    // overtime
    pub fn max_profit_1(prices: Vec<i32>) -> i32 {
        let mut ret = 0;
        for i in 0..prices.len() {
            for j in (i + 1)..prices.len() {
                ret = std::cmp::max(ret, prices[j] - prices[i]);
            }
        }
        ret
    }
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut min, mut ret) = (prices[0], 0);
        for ele in prices {
            if min > ele {
                min = ele;
            }
            ret = std::cmp::max(ret, ele - min);
        }
        ret
    }
}

fn main() {
    println!("{}", Solution::max_profit(vec![1, 2]));
}

File: leetcode/0121_maxProfit.cpp
#include "head.hpp"

using std::vector;

class Solution {
public:
  int maxProfit_timeout(vector<int>& prices) {
    int *dp = new int[prices.size()]{0};
    int max = 0;

    for (int i = 1; i < prices.size(); ++i) {
      for (int j = 0; j < i; ++j) {
        if (prices[i] > prices[j]) {
          dp[i] = std::max(dp[i], dp[j] + prices[i] - prices[j]);
          max = std::max(max, dp[i]);
        }
      }
    }
    return max;
  }

  int maxProfix(vector<int>& prices) {
    // see ./0121_max_profit.rs
  }
};

int main() {
  Solution s;
  vector<int> v1{7, 1, 5, 3, 6, 4};
  std::cout << s.maxProfit(v1) << std::endl;
  vector<int> v2{7, 6, 4, 3, 1};
  std::cout << s.maxProfit(v2) << std::endl;
  return 0;
}
